### PR TITLE
add crdt component and system

### PR DESCRIPTION
**Pull request by gonpombo8**
Add crdt system to the engine by default.

### CODE REVIEW

File: src/engine/index.ts [hunk 1]
```diff
@@ -6,13 +6,18 @@ import {
...
+import { crdtSceneSystem } from '../systems/crdt'
 
 /**
  * @alpha
  */
-export function preEngine() {
+function preEngine() {
   const entityContainer = EntityContainer()
   const componentsDefinition = new Map<number, ComponentDefinition<any>>()
+  const entitiesComponent = new Map<
```
Comment (gonpombo8) [inline src/engine/index.ts:17]: TODO: find a way to make this work. Maybe a proxy/callback to be up-to-date

File: src/engine/index.ts [hunk 2]
```diff
@@ -87,42 +103,74 @@ export function preEngine() {
...
+
+  function update(dt: number) {
+    crdtSystem.processMessages()
+
+    for (const system of engine.systems) {
+      system(dt)
+    }
+
+    // TODO: Perf tip
+    // Should we add some dirtyIteratorSet at engine level so we dont have
```
Comment (gonpombo8) [inline src/engine/index.ts:147]: TODO: Perf issues

File: src/systems/crdt/index.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import { Entity } from '../../engine/entity'
+import { createTransport } from './transport'
+import { parseKey, getKey } from './utils'
+
+/**
+ * TODO:
+ * What about when we remove the SyncComponent ? Should we notify all the clients? How ?
+ * Where do we create the transport and process the received messages?
+ */
+let i = 0
```
Comment (gonpombo8) [inline src/systems/crdt/index.ts:12]: TODO: remove, was for testing purpose only

File: src/systems/crdt/index.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+  transport.onmessage = (message: MessageEvent<string>) => {
+    console.log({ wsClient: crdtClient.getUUID(), transport: transport.url })
+    const msg: Message<Uint8Array> = JSON.parse(message.data)
+    msg.data = new Uint8Array(Object.values(msg.data))
+    messages.add(msg)
+  }
+
+  async function sendMessages(messages: Message<Uint8Array>[]) {
+    console.log({
```
Comment (gonpombo8) [inline src/systems/crdt/index.ts:27]: TODO: remove all console.logs

File: src/systems/crdt/utils.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+export function parseKey(key: string): [Entity, ClassID] {
+  const [entity, classId] = key.split('.').map(Number)
+  return [entity, classId]
+}
+
+export function getKey(entity: Entity, classId: ClassID): string {
+  return `${entity}.${classId}`
+}
+
+export enum SynchronizedEntityType {
```
Comment (gonpombo8) [inline src/systems/crdt/utils.ts:16]: TODO: next version split between diff types of transports/crdts.

File: src/systems/crdt/index.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      // CRDT outdated message. Resend this message through the wire
+      if (msg !== message) {
+        PutComponentOperation.write(entity, msg.timestamp, component, buffer)
+      } else {
+        // TODO: lean. This vs createByteBuffeR({ readingbuffer: message.data }).
+        const bb = createByteBuffer()
+        bb.writeBuffer(message.data, false)
+
+        component.upsertFromBinary(entity, bb)
+        component.clearDirty()
```
Comment (gonpombo8) [inline src/systems/crdt/index.ts:74]: TODO: How we notify to unity that this component has changed ?

File: src/systems/crdt/index.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        throw new Error(
+          'Component not found. You need to declare the components at the beginnig of the engine declaration'
+        )
+      }
+
+      // CRDT outdated message. Resend this message through the wire
+      if (msg !== message) {
+        PutComponentOperation.write(entity, msg.timestamp, component, buffer)
+      } else {
+        // TODO: lean. This vs createByteBuffeR({ readingbuffer: message.data }).
```
Comment (gonpombo8) [inline src/systems/crdt/index.ts:69]: TODO: lean. This vs createByteBuffeR({ readingbuffer: message.data }).

File: src/systems/crdt/transport.ts [hunk 8]
```diff
@@ -0,0 +1,3 @@
+export function createTransport(_type?: 'ws') {
+  return new WebSocket('ws://localhost:8000')
```
Comment (gonpombo8) [inline src/systems/crdt/transport.ts:2]: TODO: Fix transportts

**Review by leanmendoza — APPROVED** [review]

**Pull request merged**